addon/model.js

```diff
@@ -467,25 +587,165 @@ export default class MegamorphicModel extends EmberObject {
 
 MegamorphicModel.prototype.store = null;
 MegamorphicModel.prototype._internalModel = null;
+MegamorphicModel.prototype._recordData = null;
 MegamorphicModel.prototype._parentModel = null;
 MegamorphicModel.prototype._topModel = null;
 MegamorphicModel.prototype._errors = null;
+MegamorphicModel.prototype._invalidRequests = null;
+MegamorphicModel.prototype._errorRequests = null;
+MegamorphicModel.prototype._lastError = null;
 MegamorphicModel.prototype.currentState = null;
 MegamorphicModel.prototype.isError = null;
 MegamorphicModel.prototype.adapterError = null;
+MegamorphicModel.prototype._identifier = null;
 
 MegamorphicModel.relationshipsByName = new Map();
 
+/**
+    If this property is `true` the record is in the `valid` state.
+
+    A record will be in the `valid` state when the adapter did not report any
+    server-side validation failures.
+
+    @property isValid
+    @type {Boolean}
+    @readOnly
+  */
+const isValid = computed(function() {
+  if (this.get('errors.length') > 0) {
+    return false;
+  }
+  let invalidLength = this._invalidRequests.length;
+  if (invalidLength === 0) {
+    return true;
+  }
+  let invalidRequest = this._invalidRequests[invalidLength - 1];
+  if (!invalidRequest) {
+    return true;
+  } else {
+    return false;
+  }
+});
```

Comment:
Now that we have to manually implement these flags let's flesh out the unit tests at https://github.com/hjdivad/ember-m3/blob/5c6b0358e3f7e913619ba5c134fd7222b4df52dc/tests/unit/model/state-test.js#L49-L54